From 697d2ab7e70cbd22baf1da0888b4872d013d3a55 Mon Sep 17 00:00:00 2001
From: Francois Choleau <francois.choleau@st.com>
Date: Fri, 19 Apr 2024 03:55:33 -0400
Subject: [PATCH 0807/1695] drivers: stm32_rifsc: provide new API for CID
 filtering feature

Some drivers may need to disable CID filtering for specific actions.
Add a new API to know if CID filtering is enabled and to
enable/disable it

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 058bb64e36208879148b8263283ac4b6e02d2b7e drivers: stm32_rifsc: provide new API for CID filtering feature]
Signed-off-by: Francois Choleau <francois.choleau@st.com>
Change-Id: I89458ce62786e9b276512c76c54e3036f89b9fdc
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/375005
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 28 ++++++++++++++++++++++++++++
 core/include/drivers/stm32_rif.h    | 24 ++++++++++++++++++++++++
 2 files changed, 52 insertions(+)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 362323ca2..08950797a 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -553,6 +553,34 @@ TEE_Result stm32_rifsc_reconfigure_risup(unsigned int risup_id,
 	return TEE_SUCCESS;
 }
 
+bool stm32_rifsc_cid_is_enabled(unsigned int  rifc_id)
+{
+	struct io_pa_va rifsc_addr = { .pa = RIFSC_BASE };
+	vaddr_t rifsc_base = io_pa_or_va(&rifsc_addr, 1);
+	uint32_t cidcfgr = io_read32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+				     _OFST_PERX_CIDCFGR * rifc_id);
+
+	return (cidcfgr & _CIDCFGR_CFEN) == _CIDCFGR_CFEN;
+}
+
+void stm32_rifsc_cid_enable(unsigned int  rifc_id)
+{
+	struct io_pa_va rifsc_addr = { .pa = RIFSC_BASE };
+	vaddr_t rifsc_base = io_pa_or_va(&rifsc_addr, 1);
+
+	io_setbits32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+		     _OFST_PERX_CIDCFGR * rifc_id, _CIDCFGR_CFEN);
+}
+
+void stm32_rifsc_cid_disable(unsigned int  rifc_id)
+{
+	struct io_pa_va rifsc_addr = { .pa = RIFSC_BASE };
+	vaddr_t rifsc_base = io_pa_or_va(&rifsc_addr, 1);
+
+	io_clrbits32(rifsc_base + _RIFSC_RISC_PER0_CIDCFGR +
+		     _OFST_PERX_CIDCFGR * rifc_id, _CIDCFGR_CFEN);
+}
+
 TEE_Result stm32_rifsc_reconfigure_rimu(unsigned int id,
 					unsigned int master_cid,
 					bool cid_sel, bool sec, bool priv)
diff --git a/core/include/drivers/stm32_rif.h b/core/include/drivers/stm32_rif.h
index e7768a85e..65e61de52 100644
--- a/core/include/drivers/stm32_rif.h
+++ b/core/include/drivers/stm32_rif.h
@@ -135,6 +135,30 @@ TEE_Result stm32_rifsc_reconfigure_rimu(unsigned int id,
 					unsigned int master_cid,
 					bool cid_sel, bool sec, bool priv);
 
+/*
+ * Verify if CID filtering is enabled
+ *
+ * @risup_id:	ID of the RISUP to verify
+ *
+ * Returns True if CID filtering is enabled in case of success.
+ * Returns False if CID filtering is disabled
+ */
+bool stm32_rifsc_cid_is_enabled(unsigned int risup_id);
+
+/*
+ * Enable  CID filtering
+ *
+ * @risup_id:	ID of the RISUP to enable
+ */
+void stm32_rifsc_cid_enable(unsigned int risup_id);
+
+/*
+ * Disable CID filtering
+ *
+ * @risup_id:	ID of the RISUP to disable
+ */
+void stm32_rifsc_cid_disable(unsigned int risup_id);
+
 /*
  * Check every possible configuration where accessing RIF
  * is possible :
-- 
2.47.2

